base/pkg/entry.jl

```diff
@@ -242,44 +288,53 @@ function free(pkgs)
-    ispath(pkg,".git") || error("$pkg is not a git repo")
-    branch = "pinned.$(head[1:8]).tmp"
-    rslv = (head != Git.head(dir=pkg))
-    info("Creating $pkg branch $branch")
-    Git.run(`checkout -q -B $branch $head`, dir=pkg)
+    ispath(pkg,".git") || throw(PkgError("$pkg is not a git repo"))
+    rslv = !isempty(head) # no need to resolve, branch will be from HEAD
+    with(GitRepo, pkg) do repo
+        if isempty(head) # get HEAD oid
+            head = head_oid(repo)
+        end
+        branch = "pinned.$(head[1:8]).tmp"
+        info("Creating $pkg branch $branch")
+        LibGit2.create_branch(repo, branch, head)
+    end
     rslv ? resolve() : nothing
 end
-pin(pkg::AbstractString) = pin(pkg,Git.head(dir=pkg))
+pin(pkg::AbstractString) = pin(pkg, "")
 
 function pin(pkg::AbstractString, ver::VersionNumber)
-    ispath(pkg,".git") || error("$pkg is not a git repo")
-    Read.isinstalled(pkg) || error("$pkg cannot be pinned – not an installed package".tmp)
+    ispath(pkg,".git") || throw(PkgError("$pkg is not a git repo"))
+    Read.isinstalled(pkg) || throw(PkgError("$pkg cannot be pinned – not an installed package".tmp))
     avail = Read.available(pkg)
-    isempty(avail) && error("$pkg cannot be pinned – not a registered package".tmp)
-    haskey(avail,ver) || error("$pkg – $ver is not a registered version")
-    pin(pkg,avail[ver].sha1)
+    isempty(avail) && throw(PkgError("$pkg cannot be pinned – not a registered package".tmp))
+    haskey(avail,ver) || throw(PkgError("$pkg – $ver is not a registered version"))
+    pin(pkg, avail[ver].sha1)
 end
 
 function update(branch::AbstractString)
     info("Updating METADATA...")
-    cd("METADATA") do
-        if Git.branch() != branch
-            Git.dirty() && error("METADATA is dirty and not on $branch, bailing")
-            Git.attached() || error("METADATA is detached not on $branch, bailing")
-            Git.run(`fetch -q --all`)
-            Git.run(`checkout -q HEAD^0`)
-            Git.run(`branch -f $branch refs/remotes/origin/$branch`)
-            Git.run(`checkout -q $branch`)
-        end
-        # TODO: handle merge conflicts
-        Base.withenv("GIT_MERGE_AUTOEDIT"=>"no") do
-            Git.run(`pull --rebase -q`, out=DevNull)
+    with(GitRepo, "METADATA") do repo
+        with(LibGit2.head(repo)) do h
+            if LibGit2.branch(h) != branch
+                LibGit2.isdirty(repo) && throw(PkgError("METADATA is dirty and not on $branch, bailing"))
+                LibGit2.isattached(repo) || throw(PkgError("METADATA is detached not on $branch, bailing"))
+                LibGit2.fetch(repo)
+                LibGit2.checkout_head(repo)
+                LibGit2.branch!(repo, branch, track="refs/remotes/origin/$branch")
+                LibGit2.merge!(repo)
+            end
         end
+        LibGit2.fetch(repo)
+        LibGit2.merge!(repo, fast_forward=true) || LibGit2.rebase!(repo, "origin/$branch")
```

Comment:
`fast_forward=true` means this would fail any time the user has local commits in their metadata? the `pull --rebase` would be good to keep, so metadata pr's aren't full of merge commits.